Error: Cannot find module '@pnp/sp/files' -- after upgrade 2.0.8 -> 2.0.9

### Category
- [x] Bug

### Version

Please specify what version of the library you are using: [ 2.0.9 ]

Please specify what version(s) of SharePoint you are targeting: [ SP online  ]

node.js: 12.18.0

typescript: 4.0.2

OS: latest Windows 10

### Expected / Desired Behavior / Question
following code can be executed
```ts
import { SPFetchClient } from '@pnp/nodejs-commonjs';
import { sp } from '@pnp/sp-commonjs';

export function initializeSp(): void {
	sp.setup({
		sp: {
			fetchClientFactory: () => {
				return new SPFetchClient('APP_SP_SITE_URL', 'APP_LEGACY_SP_CLIENT_ID', 'APP_LEGACY_SP_CLIENT_SECRET');
			},
		},
	});
}
```

### Observed Behavior
```
C:\test\src>npm start

> nmapi@0.0.4-dev start C:\test
> npx tsc && node dist/app.js

internal/modules/cjs/loader.js:969
  throw err;
  ^

Error: Cannot find module '@pnp/sp/files'
Require stack:
- C:\test\node_modules\@pnp\nodejs-commonjs\sp-extensions\stream.js
- C:\test\node_modules\@pnp\nodejs-commonjs\sp-extensions\index.js
- C:\test\node_modules\@pnp\nodejs-commonjs\index.js
- C:\test\dist\api\news\spConnector.js
- C:\test\dist\app.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.Module._resolveFilename (C:\test\node_modules\module-alias\index.js:49:29)
    at Function.Module._load (internal/modules/cjs/loader.js:842:27)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (C:\test\node_modules\@pnp\nodejs-commonjs\sp-extensions\stream.js:6:15)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\test\\node_modules\\@pnp\\nodejs-commonjs\\sp-extensions\\stream.js',
    'C:\\test\\node_modules\\@pnp\\nodejs-commonjs\\sp-extensions\\index.js',
    'C:\\test\\node_modules\\@pnp\\nodejs-commonjs\\index.js',
    'C:\\test\\dist\\api\\news\\spConnector.js',
    'C:\\test\\dist\\app.js'
  ]
}
```

### Steps to Reproduce
On the PnPjs version 2.0.8 the procedure initializeSp() can be executed. After upgrading to v2.0.9 the code cannot be executed.

Dependencies:
```json
	"dependencies": {
		"@pnp/nodejs-commonjs": "2.0.9",
		"@pnp/sp-commonjs": "2.0.9",
		"apollo-server-koa": "~2.17.0",
		"dotenv": "~8.2.0",
		"graphql-playground-middleware-koa": "~1.6.19",
		"koa-passport": "~4.1.3",
		"koa-router": "~9.4.0",
		"module-alias": "~2.2.2",
		"passport-azure-ad": "~4.2.1"
	},
	"devDependencies": {
		"@tsconfig/node12": "~1.0.7",
		"@types/dotenv": "~8.2.0",
		"@types/koa-passport": "~4.0.2",
		"@types/koa-router": "~7.4.1",
		"@types/passport-azure-ad": "~4.0.7",
		"@typescript-eslint/eslint-plugin": "^3.10.1",
		"@typescript-eslint/parser": "^3.10.1",
		"del": "~5.1.0",
		"eslint": "^7.7.0",
		"eslint-plugin-jsdoc": "~30.3.0",
		"gulp": "~4.0.2",
		"gulp-zip": "~5.0.2",
		"husky": "~4.2.5",
		"object-assign": "~4.1.1",
		"typescript": "^4.0.2"
	}
```
tsconfig.json:
```json
{
	"extends": "@tsconfig/node12/tsconfig.json",
	"compilerOptions": {
		"esModuleInterop": false,
		"allowSyntheticDefaultImports": true,
		"incremental": true,
		"moduleResolution": "node",
		"listEmittedFiles": true,
		"locale": "en-us",
		"outDir": "dist",
		"baseUrl": ".",
		"paths": {
			"$/*": ["src/*"],
		},
		"sourceMap": true,
	},
}
```

P.S. excluding 'module-alias' from dependencies does not solve the problem:
```
C:\test>npm start

> nmapi@0.0.4-dev start C:\test
> npx tsc && node dist/app.js

internal/modules/cjs/loader.js:969
  throw err;
  ^

Error: Cannot find module '@pnp/sp/files'
Require stack:
- C:\test\node_modules\@pnp\nodejs-commonjs\sp-extensions\stream.js
- C:\test\node_modules\@pnp\nodejs-commonjs\sp-extensions\index.js
- C:\test\node_modules\@pnp\nodejs-commonjs\index.js
- C:\test\dist\api\news\spConnector.js
- C:\test\dist\app.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.Module._load (internal/modules/cjs/loader.js:842:27)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (C:\test\node_modules\@pnp\nodejs-commonjs\sp-extensions\stream.js:6:15)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\test\\node_modules\\@pnp\\nodejs-commonjs\\sp-extensions\\stream.js',
    'C:\\test\\node_modules\\@pnp\\nodejs-commonjs\\sp-extensions\\index.js',
    'C:\\test\\node_modules\\@pnp\\nodejs-commonjs\\index.js',
    'C:\\test\\dist\\api\\news\\spConnector.js',
    'C:\\test\\dist\\app.js'
  ]
}
```

--- a/buildsystem-config.ts
+++ b/buildsystem-config.ts
@@ -75,7 +75,7 @@ export default <ConfigCollection>[
                         if (/\.js$|\.d\.ts$/i.test(file.path)) {
 
                             const content: string = file.contents.toString("utf8");
-                            file.contents = Buffer.from(content.replace(/"\@pnp\/(\w*?)"/ig, `"@pnp/$1-commonjs"`));
+                            file.contents = Buffer.from(content.replace(/"\@pnp\/(\w*?)([\/|"])/ig, `"@pnp/$1-commonjs$2`));
                         }
 
                         cb(null, file);

--- a/docs/nodejs/sp-extensions.md
+++ b/docs/nodejs/sp-extensions.md
@@ -2,6 +2,8 @@
 
 By importing anything from the @pnp/nodejs library you automatically get nodejs specific extension methods added into the sp fluent api. This article describes them.
 
+> These examples use the *-commonjs version of the libraries as they target node, you can [read more about the differences](../nodejs-support.md).
+
 
 ## IFile.getStream
 
@@ -9,7 +11,7 @@ Allows you to read a response body as a nodejs PassThrough stream.
 
 ```TypeScript
 // by importing the the library the node specific extensions are automatically applied
-import { SPFetchClient } from "@pnp/nodejs-commonjs";
+import { SPFetchClient, SPNS } from "@pnp/nodejs-commonjs";
 import { sp } from "@pnp/sp-commonjs";
 
 sp.setup({
@@ -21,7 +23,7 @@ sp.setup({
 });
 
 // get the stream
-const streamResult = await sp.web.getFileByServerRelativeUrl("/sites/dev/file.txt").getStream();
+const streamResult: SPNS.IResponseBodyStream = await sp.web.getFileByServerRelativeUrl("/sites/dev/file.txt").getStream();
 
 // see if we have a known length
 console.log(streamResult.knownLength);

--- a/package.json
+++ b/package.json
@@ -6,7 +6,7 @@
   "devDependencies": {
     "@istanbuljs/nyc-config-typescript": "^0.1.3",
     "@microsoft/microsoft-graph-types": "^1.17.0",
-    "@pnp/buildsystem": "^2.2.2",
+    "@pnp/buildsystem": "^2.2.3",
     "@types/adal-angular": "^1.0.1",
     "@types/chai": "^4.1.7",
     "@types/chai-as-promised": "7.1.0",

--- a/packages/nodejs/sp-extensions/stream.ts
+++ b/packages/nodejs/sp-extensions/stream.ts
@@ -21,12 +21,9 @@ extendFactory(File, {
     },
 });
 
-// for extensions to correctly appear in intellisense we need to extend the interface
-// to do this we extend the modules and need to append the /types to the normal import path
-// this has to do with where the file we are extending is located
 declare module "@pnp/sp/files/types" {
     /**
-     * Returns the instance wrapped by the invokable proxy
+     * Gets a PassThrough stream representing the file
      */
     interface IFile {
         getStream(): Promise<IResponseBodyStream>;

--- a/packages/odata/invokable-extensions.ts
+++ b/packages/odata/invokable-extensions.ts
@@ -1,4 +1,4 @@
-import { isFunc, isArray, ITypedHash } from "@pnp/common";
+import { isFunc, isArray, ITypedHash, getGUID } from "@pnp/common";
 
 export type ValidProxyMethods = "apply" | "get" | "has" | "set";
 
@@ -7,9 +7,10 @@ export type ExtensionDelegateType<T extends object> = { (op: string, target: T, 
 export type ExtensionType<T extends object = {}> = Pick<ProxyHandler<T>, ValidProxyMethods> | ExtensionDelegateType<T> | ITypedHash<any>;
 
 let _enableExtensions = false;
-const globaExtensions: ExtensionType[] = [];
+const globalExtensions: ExtensionType[] = [];
+const factoryExtensions: Map<string, ExtensionType[]> = new Map<string, ExtensionType[]>();
 
-const ObjExtensionsSym = Symbol("__extensions");
+const ObjExtensionsSym = Symbol.for("43f7a601");
 
 /**
  * Creates global extensions across all invokable objects
@@ -19,7 +20,7 @@ const ObjExtensionsSym = Symbol("__extensions");
 export const extendGlobal = (e: ExtensionType | ExtensionType[]) => {
 
     _enableExtensions = true;
-    extendCol(globaExtensions, e);
+    extendCol(globalExtensions, e);
 };
 
 /**
@@ -51,11 +52,23 @@ export const extendFactory = <T extends (...args: any[]) => any>(factory: T, ext
 
     _enableExtensions = true;
 
-    if ((<any>factory).__proto__[ObjExtensionsSym] === undefined) {
-        (<any>factory).__proto__[ObjExtensionsSym] = [];
+    // factoryExtensions
+    const proto = Reflect.getPrototypeOf(factory);
+
+    if (!Reflect.has(proto, ObjExtensionsSym)) {
+
+        Reflect.defineProperty(proto, ObjExtensionsSym, {
+            value: getGUID(),
+        });
     }
 
-    extendCol((<any>factory).__proto__[ObjExtensionsSym], extensions);
+    const key = proto[ObjExtensionsSym];
+
+    if (!factoryExtensions.has(key)) {
+        factoryExtensions.set(key, []);
+    }
+
+    extendCol(factoryExtensions.get(key), extensions);
 };
 
 function extendCol(a: ExtensionType[], e: ExtensionType | ExtensionType[]) {
@@ -72,7 +85,7 @@ function extendCol(a: ExtensionType[], e: ExtensionType | ExtensionType[]) {
  * Clears all global extensions
  */
 export const clearGlobalExtensions = () => {
-    globaExtensions.length = 0;
+    globalExtensions.length = 0;
 };
 
 /**
@@ -99,8 +112,13 @@ export const applyFactoryExtensions = <T extends object = {}>(factory: (args: an
 
     let o = factory(args);
 
-    if ((<any>factory).__proto__[ObjExtensionsSym]) {
-        o = extendObj(o, (<any>factory).__proto__[ObjExtensionsSym]);
+    const proto = Reflect.getPrototypeOf(factory);
+
+    if (Reflect.has(proto, ObjExtensionsSym)) {
+
+        const extensions = factoryExtensions.get(Reflect.get(proto, ObjExtensionsSym));
+
+        o = extendObj(o, extensions);
     }
 
     return o;
@@ -118,7 +136,7 @@ export function extensionOrDefault(op: ValidProxyMethods, or: (...args: any[]) =
         }
 
         // second we need to process any global extensions
-        extensions.push(...globaExtensions);
+        extensions.push(...globalExtensions);
 
         for (let i = 0; i < extensions.length; i++) {
             const extension = extensions[i];

--- a/packages/sp/batch.ts
+++ b/packages/sp/batch.ts
@@ -1,4 +1,4 @@
-import { Batch } from "@pnp/odata";
+import { Batch, HttpRequestError } from "@pnp/odata";
 import { getGUID, isUrlAbsolute, combine, mergeHeaders, hOP } from "@pnp/common";
 import { Logger, LogLevel } from "@pnp/logging";
 import { SPHttpClient } from "./sphttpclient";
@@ -207,7 +207,14 @@ export class SPBatch extends Batch {
         Logger.write(`[${this.batchId}] (${(new Date()).getTime()}) Sending batch request.`, LogLevel.Info);
 
         const fetchResponse = await client.fetch(combine(absoluteRequestUrl, "/_api/$batch"), batchOptions);
-        const text = await fetchResponse.text();
+
+        if (!fetchResponse.ok) {
+            // the entire batch resulted in an error and we need to handle that better #1356
+            // things consistently with the rest of the http errors
+            throw (await HttpRequestError.init(fetchResponse));
+        }
+
+        const text = await fetchResponse.clone().text();
         const responses = SPBatch.ParseResponse(text);
 
         if (responses.length !== this.requests.length) {

--- a/test/nodejs/sp-extenstions.ts
+++ b/test/nodejs/sp-extenstions.ts
@@ -25,7 +25,7 @@ describe("nodejs - sp-extensions", () => {
             // tslint:disable-next-line: no-unused-expression
             expect(stream).to.not.be.null;
 
-            expect(stream.knownLength).to.not.be.null.and.be.greaterThan(0);
+            expect(stream.knownLength).to.be.greaterThan(0);
 
             // tslint:disable-next-line: no-unused-expression
             expect(stream.body).to.not.be.null;

--- a/tools/buildsystem/package.json
+++ b/tools/buildsystem/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@pnp/buildsystem",
-  "version": "2.2.2",
+  "version": "2.2.3",
   "bin": {
     "pnpbuild": "./bin/buildsystem.js"
   },

--- a/tools/buildsystem/src/tasks/package/copy-target-files.ts
+++ b/tools/buildsystem/src/tasks/package/copy-target-files.ts
@@ -28,7 +28,7 @@ export function createCopyTargetFiles(targetOverride = "", subDir = "", transfor
             await new Promise((res, rej) => {
 
                 pump([
-                    src(["./**/*.d.ts", "./**/*.js"], {
+                    src(["./**/*.d.ts", "./**/*.js", "./**/*.js.map", "./**/*.d.ts.map"], {
                         cwd: resolve(buildOutDir, dir),
                     }),
                     ...transforms.map(t => obj(t)),

